Experimental skip-ci support for github

tume opened the pull request:
My first lines of Go code so apologies if it is totally out of place :)

I was trying to implement this in the common code as suggested here: https://github.com/bitrise-io/bitrise-webhooks/issues/2 but couldn't realize how to access the commit message from there so decided to put in the TransformRequest instead which seemed like good place because it already has ShouldSkip.

Would like to hear if this is good way to do it or if it should be placed somewhere else? 
Also this only implements it for the github push event so would need to implement it other places also where needed...

viktorbenei commented:
Hi,

Thanks for the PR!

A while back I was thinking about where this should be implemented, right now it would be here https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L178

At that point you have access to the converted "commit" message, regardless of which service triggered the hook.

The only issue is, how it should respond to the call. E.g. in case the webhook includes more than one commit, one should be skipped but the other one should trigger a build (BitBucket sends batched commits in some cases for example). You can't just return a single "OK, but skipped" like this https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L143 .

I think this should still be a "success" (https://github.com/bitrise-io/bitrise-webhooks/blob/master/service/hook/endpoint.go#L181), but with a special message like the "skipped" one.

WDYT?

viktorbenei commented:
BTW the error in the PR is that you did not handle the `err` returned by the `regex.Compile`:

> service/hook/github/github.go:86:5	r, _ := regexp.Compile(`(?i)\[skip ci\]|\[ci skip\]`)

I'd also suggest you to not to use a regex, a simple [strings.Contains](https://golang.org/pkg/strings/#Contains) should be enough, but of course `regexp` is fine too ;)

viktorbenei commented:
Additional note for testing: you should install the [Bitrise CLI](https://www.bitrise.io/cli), then you can run `bitrise run test` in the code's root directory (where the `bitrise.yml` file is) to run the full test suite (the PR check runs the same tests).

tume commented:
Thanks for your comments that sounds like a better approach :). Didn't even realize that you can also receive commits in a batch. I can try redoing my work with that information and see where that leads to.

viktorbenei commented:
Awesome, let us know if we can help with anything! ;)

tume commented:
Another try :). 

Had to modify a bit more than wanted in the beginning. Idea was to add new response specific for the skip cases. Not sure if that is what you were thinking but at least this way it should not be problem if the bitbucket sends them in batches.

No tests for this and it now applies the same logic for pull request which might not be desirable.

viktorbenei commented on service/hook/endpoint.go:
```diff
@@ -171,10 +172,22 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
 	respondWith := hookCommon.TransformResponseInputModel{
 		Errors:                  []string{},
 		SuccessTriggerResponses: []bitriseapi.TriggerAPIResponseModel{},
+		SkipResponses:           []bitriseapi.SkipAPIResponseModel{},
```
to conform to the naming of the other related params rename this to `SkippedTriggerResponses`

viktorbenei commented on service/hook/default_response_provider.go:
```diff
@@ -27,13 +27,19 @@ type SuccessRespModel struct {
 	httpStatusCode := 201
+
+	if len(input.SuccessTriggerResponses) == 0 && len(input.SkipResponses) > 0 {
+		httpStatusCode = 200
```
good idea - please update the README's related section about the possible HTTP status codes the server can return

viktorbenei replied:
actually.. shouldn't we return with a `400` code instead? to make it stand out in the lists, e.g. when you check the webhook history on GitHub and you try to see why a given commit did not start a build?

On the other hand a skip probably shouldn't be `400`, unless there actually was an error somewhere.

WDYT?

tume replied:
This was a bit tricky for me also. I would still consider it to be success when skipped instead of a failure. At least 400 sounds weird for me because it usually means that there were some problems with the request. 

Might be that it should be something else than 200 but didn't really find any better alternative when quickly looking at the available status codes.

viktorbenei replied:
I agree, most likely `200` is a good code for this case; when the build is actually started the code is `201` anyway, that should be enough indication I think

viktorbenei commented on bitriseapi/bitriseapi.go:
```diff
@@ -39,6 +39,13 @@ type TriggerAPIResponseModel struct {
 
+// SkipAPIResponseModel ...
+type SkipAPIResponseModel struct {
+	CommitHash    string `json:"commit_hash"`
```
I'm not sure whether these params are important at all - the webhook's "request" info should include these, but might be a good idea.

What I'd definitely add is:

```
Message           string `json:"message"`
```

Which would include the description of why it was skipped

viktorbenei commented on service/hook/default_response_provider.go:
```diff
@@ -43,6 +49,7 @@ func (hp DefaultResponseProvider) TransformResponse(input hookCommon.TransformRe
 			Errors:                  input.Errors,
 			SuccessTriggerResponses: input.SuccessTriggerResponses,
 			FailedTriggerResponses:  input.FailedTriggerResponses,
+			SkipResponses:  input.SkipResponses,
```
please run `go fmt service/...` ;)

viktorbenei commented on service/hook/common/common.go:
```diff
@@ -43,6 +43,8 @@ type TransformResponseModel struct {
 type TransformResponseInputModel struct {
 	// Errors include the errors if the build could not trigger
 	Errors []string
+	SkipResponses        []bitriseapi.SkipAPIResponseModel
```
to conform to the naming of the other related params rename this to `SkippedTriggerResponses`

viktorbenei commented:
LGTM, added a couple of things I think should be changed, but I agree with the solution / where and how you implemented the skip functionality ;)

tume commented:
- Renamed SkipResponses -> SkippedTriggerResponses. 
- Fixed styling issues.
- Added message to SkipAPIResponseModel just not sure what text to return :) and left commitMessage/has there but can remove if not needed.

viktorbenei commented on service/hook/endpoint.go:
```diff
@@ -171,10 +172,23 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
+
+			if strings.Contains(commitMessage, "[skip ci]") || strings.Contains(commitMessage, "[ci skip]") {
+				respondWith.SkippedTriggerResponses = append(respondWith.SkippedTriggerResponses, bitriseapi.SkipAPIResponseModel{
+					Message:       "Build skipped by request",
```
Build skipped because the commit message included a skip ci keyword ([skip ci] or [ci skip]).

tume replied:
Changed

viktorbenei commented:
LGTM, just added a comment for the skip response message.

One more thing what would be awesome is to add some tests, as well as adding this to the README. Feel free to add yourself as a contributor too ;)

tume commented:
If it is looking good otherwise I can try writing couple of test cases during the weekend.

viktorbenei commented:
It looks perfect, we'll merge it early next week. If you have the time to write tests it would be awesome, otherwise we can merge it as-it-is and write the tests after the merge.

viktorbenei commented on service/hook/default_response_provider.go:
```diff
@@ -27,13 +27,19 @@ type SuccessRespModel struct {
 // TransformResponseModel ...
 type TransformResponseModel struct {
 	Errors                  []string                             `json:"errors,omitempty"`
+	SkippedTriggerResponses []bitriseapi.SkipAPIResponseModel    `json:"skipped,omitempty"`
```
`json:"skipped,omitempty"` should be `json:"skipped_responses,omitempty"` - to conform to the other response names

viktorbenei replied:
and move it to be the last one, after `FailedTriggerResponses`

viktorbenei commented on service/hook/common/common.go:
```diff
@@ -42,7 +42,9 @@ type TransformResponseModel struct {
 	// Errors include the errors if the build could not trigger
-	Errors []string
+	Errors                  []string
+	SkippedTriggerResponses []bitriseapi.SkipAPIResponseModel
```
Add comment - like the one `SuccessTriggerResponses` has and move it to be the last one, after `FailedTriggerResponses`

viktorbenei commented on service/hook/endpoint.go:
```diff
@@ -171,10 +172,23 @@ func HTTPHandler(w http.ResponseWriter, r *http.Request) {
 		for _, aBuildTriggerParam := range hookTransformResult.TriggerAPIParams {
+			commitMessage := aBuildTriggerParam.BuildParams.CommitMessage
+
+			if strings.Contains(commitMessage, "[skip ci]") || strings.Contains(commitMessage, "[ci skip]") {
```
move this into a function - and write tests for it!

write both "yes, should skip" and "no, this should not be skipped" tests.

Example commit messages to test against:

* `this should be [ci skip]ped`
* `this will not be [CI SKIP]ped`
* `this won't be skipped either: [ ci  skip ]`
* `[skip ci] will skip`
* ...

We use unit tests to document the code, so you should try to add tests to demonstrate both what will be skipped and what won't be skipped, including possible edge cases (e.g. case sensitivity).

tume commented:
Sorry have been a bit busy with other work but will try to fix those tomorrow.

viktorbenei commented:
Sure, no problem at all, let us know if we can help with anything!

viktorbenei commented:
Let us know if you need any help!

viktorbenei commented:
For reference, the related UserVoice feature request https://bitrise.uservoice.com/forums/235233-general/suggestions/10519956-allow-for-configurable-commit-message-pattern-s-t

(we don't plan to make the pattern configurable, at least for now)

callmephilip commented:
really looking forward for this being merged in 👍

nobre84 commented:
Having this will be awesome! 👍

tume commented:
Got back from vacation today so can continue where I left last time :)

Fixed code style/comments that you mentioned and will look into tests next after I first get some idea how the tests are done currently.

viktorbenei commented:
Awesome to hear @tume , just wanted to merge & finish it myself, but it'd be awesome if you could do it instead!

One note: we discussed about whether `[ci skip]` / `[skip ci]` should be searched in the whole commit text, or only in the first line, and we believe that the best solution would be to search for it only in the first line of the commit text.

The reason for this is that e.g. when you do a "squash and merge" here on GitHub, the first line of the commit (title) will be the title of the Pull Request, and the content (other lines of the commit text / commit body) will be a list of all the commits in the PR. It's completely possible that you pushed a commit with "ci skip" during the development of the PR, but of course you don't want to skip the build after the PR is merged.

If we'd search for the pattern in the whole commit text then you should make sure when you squash and merge that you remove it from everywhere in the commit message, while if we only search for the pattern in the first line / title of the commit then it's way easier to spot if you have the pattern in it.

What do you think about this @tume ?

viktorbenei commented:
For reference, if I'd do a "squash and merge" on this PR right now, without changing the texts auto generated by GitHub, the full commit text (title + body) would look like:

```
Experimental skip-ci support for github (#8)

* Generic skip ci support

* Review fixes

* Changed skipped message

* Improve code comments

```

bootstraponline commented:
> One note: we discussed about whether [ci skip] / [skip ci] should be searched in the whole commit text, or only in the first line, and we believe that the best solution would be to search for it only in the first line of the commit text.

Pls no. I put `[ci skip]` at the end of the commit text. This is how it's done on Travis. 

> Commits that have [ci skip] or [skip ci] anywhere in the commit messages are ignored by Travis CI.
>
> https://docs.travis-ci.com/user/customizing-the-build#Skipping-a-build

viktorbenei commented:
@bootstraponline why's putting it into the first line an issue? That would solve the problem/use case I mentioned:

> The reason for this is that e.g. when you do a "squash and merge" here on GitHub, the first line of the commit (title) will be the title of the Pull Request, and the content (other lines of the commit text / commit body) will be a list of all the commits in the PR. It's completely possible that you pushed a commit with "ci skip" during the development of the PR, but of course you don't want to skip the build after the PR is merged.

bootstraponline commented:
> @bootstraponline why's putting it into the first line an issue? That would solve the problem/use case I mentioned:

1) This is not a real issue, I haven't seen anyone petitioning Travis to adopt a similar fix.
2) The first line is reserved for use by humans, cluttering it up with messages to the CI service is not a good practice.

My typical message is like this:

````
Update selenium-webdriver dependency to 2.48

[ci skip]
````

viktorbenei commented:
Makes sense, I'm just worried that it's too easy to miss a `[ci skip]` in a merge's commit message, auto generated by a tool - we do this a lot actually, when the commit history is clean and there's no point in changing the commit texts.

bootstraponline commented:
Let's wait to see if people actually experience this as a legitimate issue.
I think the use case I described is the norm. There should be a very good
reason to break that expectation.

On Thu, Sep 8, 2016, 5:05 PM Viktor Benei <notifications@github.com> wrote:

> Makes sense, I'm just worried that it's too easy to miss a [ci skip] in a
> merge's commit message, auto generated by a tool - we do this a lot
> actually, when the commit history is clean and there's no point in changing
> the commit texts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitrise-io/bitrise-webhooks/pull/8#issuecomment-245740733>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABHmQUJJJ90iF90ixio7LC__6Bq4rn3tks5qoHicgaJpZM4IvMdE>
> .
>

viktorbenei commented:
I agree, new features are always subject for revisions anyway if it turns out a use case was missed.

viktorbenei commented:
btw @bootstraponline what about first __and__ last line? Would that work?
I'm personally still not sure whether the middle of the commit text should be checked for the pattern, it's too easy to miss IMO

bootstraponline commented:
> btw @bootstraponline what about first and last line?  Would that work?

No, that's too magical. If CI skip is present then bitrise shouldn't ignore it.

```
Add some numbers

[ci skip]
Fix more stuff
```

viktorbenei commented:
@bootstraponline re

```
Add some numbers

[ci skip]
Fix more stuff
```

IMO this commit should be built, as you most likely left `[ci skip]` in it by accident and not by purpose

bootstraponline commented:
> IMO this commit should be built, as you most likely left [ci skip] in it by accident and not by purpose

Uh, no. It's following the Travis convention which has set the standard. This "by accident" use case doesn't exist.

Is there any CI provider which intentionally disregards users when they explicitly mark a commit as skip?

viktorbenei commented:
I definitely agree with you @bootstraponline , in general, but given the time we spend in support.. You would be amazed how many times people have issues like this (when they forget to change something and just click, "next next next" style)

bootstraponline commented:
I would expect you'd get more support issues by being the only vendor in this space to have a non-standard ci skip behavior. They seem to manage support despite having a much larger customer base.

> CircleCI won’t run the build if we find [ci skip] or [skip ci] anywhere in the commit message of the head commit.
>
> https://circleci.com/docs/skip-a-build/

--

> Commits that have [ci skip] or [skip ci] anywhere in the commit messages are ignored by Travis CI.
>
> https://docs.travis-ci.com/user/customizing-the-build/

--

> [skip ci] to the commit message of the last commit before you push and that push will be ignored. 
>
> https://documentation.codeship.com/continuous-integration/skipping-builds/

viktorbenei commented:
@bootstraponline I'm completely aware how it is in other CI/CD services, but that doesn't mean that there's no better solution, and that we can't discuss about possible use cases where that solution would not work or could be improved

bootstraponline commented:
I just think something needs to be implemented and then we collect data based on the reality. There's only so much you can figure out without running code.

viktorbenei commented:
@bootstraponline that I completely agree with ;)

And, why we discuss these last details here is exactly because we plan to merge and release this _real soon_ ;)

callmephilip commented:
Really looking forward to this one! 🎈

viktorbenei commented:
@tume added you as a contributor into the README -> https://github.com/bitrise-io/bitrise-webhooks/pull/19/commits/32ce12875f03f88c453872944131f88d028c462c

if you'd want to change that please let me know!

viktorbenei commented:
Skip CI is now in production, if you use the hooks.bitrise.io server for your webhooks (the default) you can now use `[skip ci]` and `[ci skip]`!

Huge thanks to @tume for the PR, and for everybody else who contributed to the discussion here! :rocket: :tada:

nobre84 commented:
Great stuff ! 🎉 🎉

callmephilip commented:
yay! 🎆

bootstraponline commented:
awesome! :shipit:
